music/src/core/player.ts

```diff
@@ -39,21 +39,76 @@ function compareQuantizedNotes(a: NoteSequence.INote, b: NoteSequence.INote) {
   return 1;
 }
 
+/**
+ * An abstract base class for providing arbitrary callbacks for each note
+ * played.
+ */
+export abstract class BasePlayerCallback {
+
+  /* Will be called for each time/note pair in a sequence being played.
+   */
+  abstract run(n: NoteSequence.INote): void;
+
+  /*  Will be called when a sequence is stopped.
+   */
+  abstract stop(): void;
+}
+
 /**
  * Abstract base class for a `NoteSequence` player based on Tone.js.
  */
 export abstract class BasePlayer {
-  private currentPart: any;  // tslint:disable-line:no-any
-  private scheduledStop: number;
+  protected currentPart: any;  // tslint:disable-line:no-any
+  protected scheduledStop: number;
+  protected playClick: boolean;
+  protected callbackObject: BasePlayerCallback;
 
   protected abstract playNote(time: number, note: NoteSequence.INote): void;
 
-  constructor() {
+  /**
+   *   `BasePlayer` constructor.
+   *
+   *   @param playClick A boolean, determines whether the click will be played.
+   *   @param callbackObject An optional BasePlayerCallback, specifies an
+   *     object that contains run() and stop() methods to invode during
+   *     playback.
+   */
+  constructor(playClick=true, callbackObject?: BasePlayerCallback) {
```

Comment:
default should be false for playclick

Reply:
Perhaps it should be clickSignature instead of playClick, which may be udnefined to disable. Perhaps we only support 4/4 at first, but this will make it easier to extend later.

Reply:
made false by default.
do NoteSequences normally have time signature defined? if so, it probably makes sense to just use that rather than allowing arbitrary time signatures that don't necessarily correspond to the note sequence.
thoughts?